Initial feature specification for horizontal inference

stereotype441 opened the pull request:
Initial attempt at a spec to address #731.  I believe this correctly reflects what I've prototyped.

This is my first feature spec, so I would not be surprised if I'm missing some important details.  Let me know what you think!

srawlins commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+### Topological sort
+
+In the algorithm above, the "topological sort" in step 3 works as follows.
```
I believe this is only referenced in step 4.

stereotype441 replied:
Fixed, thank you!

eernstg commented:
Sounds good! I'll return to this after the Easter days.

stereotype441 commented:
> Sounds good! I'll return to this after the Easter days.

@eernstg Friendly ping!  I hope you had a good holiday 😃

eernstg commented:
Very nice Easter, thanks! Had to work on other stuff today (such as slides for the Thursday 'Deep Dive', and an interview), but I'll get to this very soon.

eernstg commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+placed in the first stage) is critically important&mdash;it guarantees that the
+only arguments that will be type inferred out of order are function literals.
+
+### Why do we defer all function literals?
```
This is the part that I think ought to be handled as a separate type inference improvement. Regardless of dependencies and argument ordering, it should be safe to postpone the introduction of the write-capturing effect of function literal arguments (that is, function literals that occur in an actual argument list and are not invoked during argument evaluation) until after the evaluation of all arguments.

If we do that separately then we keep things separate just because that is possible (which is an overall simplification), and then we don't need to consider a subdivision of the work in each stage (which turns out to be stage 0 only anyway, because that's the only one which can have both function literal arguments and other arguments).

I don't think it actually matters, but one part of the simplification would be that we could have, say, 3 stages, and then we could introduce some write-capture effects after stage 0, some other ones after stage 1, and finally some more after stage 2.

If we handle this aspect separately then we'd just gather the write-capture effects along the way and perform all of them after having handled the entire argument list.

stereotype441 replied:
As I mentioned earlier, I prefer keeping the two improvements in the same proposal, both because the implementation is nearly complete (so there's no benefit in terms of implementation effort to splitting it up), and because the current behavior is poorly specified (so we can't really rewrite this into two separate smaller improvements without making reference to a specification that doesn't exist).

eernstg replied:
OK!

eernstg reviewed (commented):
Finally got around to finish this review, sorry about the delay!

There are lots of comments! I hope the explanations are meaningful and actionable. One reason why there are so many comments is that the specification of inference is at a somewhat early stage. We haven't integrated any of these things into the language specification, and some parts of 'inference.md' haven't been written (or at least finalized) at this point, so I tend to ask for extra information in a rather large number of situations.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
+
```
Good points.  I've introduced a "terminology" section and I've tried to capture all the concepts I've used that are defined elsewhere.

eernstg replied:
Thanks! It gets a bit heavy with all these explicit definitions, but I think it's very important to have the definitions somewhere. The language specification avoids being dominated by definitions only because it's longer, so we have the benefits of scale by defining a concept, say, on page 59, and then using it everywhere else in the document, relying on the index to find the definition if needed. So this document could be abbreviated somewhat by omitting some of the 4-5 line definitions, and just put a link to an existing definition in some other document at the first usage of the given word/phrase, but I don't think it's important to put a lot of time into that kind of improvement.

The main point is that it's absolutely great that we have definitions for a lot of things that were previously a bit harder to resolve!

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+## Type inference algorithm for invocations
+
+The full algorithm for type inference of invocations is expanded as follows.
```
I see where you're coming from, however I prefer keeping the two improvements in the same proposal, for two reasons. 
 (1) Both improvements have already been implemented in the analyzer, and a full implementation in the front end should be finished soon, so separating the two proposals wouldn't actually give us any benefit in terms of implementation flexibility.  (2) Since the algorithm for type inference of invocations (as it exists in Dart 2.18) is not currently specified anywhere, I believe that breaking the proposal into two wouldn't actually be any easier for readers to understand, because although the two individual proposals would be shorter, they would both be written as deltas against an implicit and unwritten base algorithm.

eernstg replied:
OK!

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+This algorithm is used for constructor invocations (including those in
+annotations, super-constructor calls, and constructor redirects), method
+invocations (including super calls), and invocations of function-typed
+expressions.
```
I've cleaned this up by making explicit reference to syntactic constructs using the names that are used in the spec.

eernstg replied:
Thanks!

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+1. Initial constraints: Create some constraints on type parameters by trying to
+   match the return type of the invocation target as a subtype of the incoming
+   type context.  (For a constructor invocation, the return type of the
```
Correct.  I've clarified this in the "terminology" section.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+   match the return type of the invocation target as a subtype of the incoming
+   type context.  (For a constructor invocation, the return type of the
+   invocation target is considered the raw uninstantiated type of the class
+   enclosing the constructor declaration.)
```
Code in the front end uses the term "raw type" to refer to a generic type in which each type argument is slot is filled by a type variable whose name matches that of the corresponding generic type parameter.  So, in front end terminology, `C<X, Y>` is the "raw type" associated with the class `C`.  I've dropped the use of "raw type" in my terminology clean-up; now I simply refer to the "function type of the corresponding constructor", which is defined in the spec in the way that I need (e.g. for `class C<X extends B, Y> { C(); }`, the function type of the constructor is defined to be `C<X, Y> Function<X extends B, Y>()`).

eernstg replied:
OK, good to know, and thanks for the adjustment!

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+2. Downwards inference: partially solve the set of type constraints accumulated
+   in step 1, to produce a preliminary mapping of type parameters to type
+   schemas.
```
> Sounds like 'partially solve' is related to ['Constraint solution for a type variable'](https://github.com/dart-lang/language/blob/master/resources/type-system/inference.md#constraint-solution-for-a-type-variable). If so, please mention it.

Done.

> We could have a sentence here saying 'Proceed to step 3 if "legacy", and to step 4 if "new"', in order to clarify how we could end up being at step 4 (the current approach is kind of "COMEFROM"-ish, which is harder to handle than "GOTO" ;-). But I'd recommend avoiding the description of both algorithms at the same time anyway—see comment on line 66.

Good point.  I've gone ahead with your suggestion and just described the "new" algorithm.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+   in step 1, to produce a preliminary mapping of type parameters to type
+   schemas.
+   
+3. Visit arguments (legacy): If experiment flag `inference-update-1` is not
```
Agreed.  I've reworked the document mostly according to your suggestions.  Rather than include a `3-current` I've simply added a non-normative section near the end called "Differences from language version 2.18" that describes in prose how the current behavior differs.  I believe the difference is clear enough that it doesn't need to be spelled out formally.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+   * Recursively infer all arguments selected for this stage that are not
+     function literals (see [what constitutes a function
+     literal?](#what-constitutes-a-function-literal), in source order.  (Note
+     that an invariant of the topological sort is that non-function literals are
```
Thanks for the suggestion.  I've updated the nomenclature mostly to "argument that is a function literal expression" and "argument that is not a function literal expression".  In one place in non-normative text, that would have been awkward so I went with "non-(function literal) argument".

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+     literal?](#what-constitutes-a-function-literal), in source order.  (Note
+     that an invariant of the topological sort is that non-function literals are
+     always placed in the first stage).  The logic for obtaining type contexts
+     and for creating additional constraints is the same as in step 3.
```
Acknowledged.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+     and for creating additional constraints is the same as in step 3.
+
+   * Recursively infer all arguments selected for this stage that are function
+     literals, in source order.
```
Heh, this is funny because an earlier (doc-based) version of the spec, I did not specify the order of processing function literal arguments, and Lasse suggested that we ought to specify it anyhow (even though it _shouldn't_ matter), because if there's any chance that we're wrong and it actually _does_ matter in some subtle way, we want to make sure the user sees consistent behavior.

I was swayed by that argument and decided to go ahead and specify the order.  I've added normative text explaining that we don't believe the order should matter, but we're specifying it anyway to reduce the risk of unpredictable behavior.

eernstg replied:
Heh. ;-)

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+     to step 5.
+
+5. Upwards inference: solve the set of type constraints accumulated so far, to
+   produce a final mapping of type parameters to types.  Check that each type is
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+5. Upwards inference: solve the set of type constraints accumulated so far, to
+   produce a final mapping of type parameters to types.  Check that each type is
+   a subtype of the bound of its corresponding type parameter.
```
I've added the term "(possibly grounded) constraint solution for a set of type variables, considering bounds" to the "terminology" section to explain how this works.  The short story is that every time we solve the constraints, if we're about to choose a solution that doesn't refer to the unknown type, then at the last minute we throw in an extra constraint to ensure that the bound is met.  The full details of how this works are tricky (especially in the case where one type parameter's bound may refer to other type parameters).  AFAICT these details are not specified anywhere, but since I'm not proposing to change them in any way, and since this proposal has already gotten quite big and detailed, I decided to draw a cut line here and not delve further into the handling of bounds.

eernstg replied:
It is slightly dangerous to define two distinct concepts using metalanguage. So "(possibly grounded) constraint solution for a set of type variables, considering bounds" is really intended to mean "constraint solution for a set of type variables, considering bounds" (with one definition that follows), and "grounded constraint solution for a set of type variables, considering bounds" (with another definition). I often choose to use a few extra words in order to avoid introducing new, fancy meta-language mechanisms (and introducing them implicitly!), but it shouldn't be a big problem here.

eernstg replied:
(I kept this thread unresolved, not because I'm asking for further changes, but just because I wanted to keep the previous comment visible.)

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+(Note that step 4 is the only step that differs from the inference implemented
+in Dart today, and it only takes effect if the experiment flag
+`inference-update-1` is enabled, hence this feature is backward compatible.)
```
I wasn't really trying to claim that code that works without the flag enabled would continue to work with the flag enabled.  I was just trying to say that if the experiment flag is disabled, the implementations will continue to behave the way they've always behaved.  Which is kind of a vacuous statement anyhow, since the whole proposal is meant to only take effect when the experiment flag is enabled.  So I've just dropped this paragraph.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+following rule: there is a dependency edge from argument _A_ to argument _B_ if
+and only if the type of the invocation target is generic, and the following
+relationship exists among _A_, _B_, and at least one of the invocation target’s
+type formals _T_:
```
Good point.  Fixed.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+type formals _T_:
+
+1. _A_ is a function literal (see [what constitutes a function
+   literal?](#what-constitutes-a-function-literal))
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+2. AND the parameter in the invocation target corresponding to _A_ is function
+   typed,
+
+3. AND _T_ is a free variable in at least one of the parameters of that function
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+3. AND _T_ is a free variable in at least one of the parameters of that function
+   type,
+
+4. AND the corresponding parameter in _A_ is implicitly typed,
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+(The idea here is that we're trying to draw a dependency edge from _A_ to _B_ in
+precisely those cirumstances in which there is likely to be a benefit to
+performing a round of horizontal inference between type inferring _A_ and type
+inferring _B_).
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+type context is `int Function(int, int)`.  This results in `a` being assigned a
+static type of `int`, which is what the user intends.
+
+## Type inference algorithm for invocations
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+inferring _B_).
+
+If the type of the invocation target is `dynamic` or the `Function` class, or
+some non-generic function type, then the resulting graph has no edges.
```
Fair point.  However, since there is no existing specification for type inference of invocations, and I'm trying to supply one as part of this proposal, it seemed reasonable to include details about what happens in the case of dynamic invocations.

eernstg replied:
Thanks!

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+[guaranteed to be acyclic]: https://en.wikipedia.org/wiki/Strongly_connected_component#Definitions
+
+&emsp;(B) &lArr; (A) &rArr; (C &hArr; D)
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+the process for successive stages, until no strongly connected components remain
+in the graph.  The set of arguments we will recursively infer during each stage
+is the union of the arguments in all the strongly connected components removed
+from the graph during that stage.
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+
+In this example, that means that there will be two stages.  The first stage will
+consist of argument A, and the second stage will consist of arguments B, C, and
+D.
```
Done.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+always placed in the first stage, is guaranteed by the fact that dependency
+analysis only draws an edge from _A_ to _B_ if _A_ is a function literal.)
+
+### What constitutes a function literal?
```
Done.  I decided I wasn't comfortable with the phrase "function literal argument" because the recursive nature of the definition means that we would be calling some things "function literal arguments" that are not, in fact, arguments at all, so I instead chose the phrase "function literal expression".

eernstg replied:
Works!

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,290 @@
+re-ordered within the invocation, because the act of _evaluating_ a function
+literal doesn't actually have any runtime effect.  The runtime effect doesn't
+happen until the function literal is _called_, and the earliest this can
+possibly happen until after all the invocation arguments have been evaluated.
```
Yes, that's true.  Flow analysis understands that, upon entry to the top of the `while` loop, any write captures that occur inside the loop may already have occurred (because the loop may execute more than once), so it considers any write captures inside the loop itself to take effect at the top of the loop.

stereotype441 commented:
Thanks for the review, @eernstg!  I believe I've addressed all of your comments (or explained why I think things should stay as they are).  Please take another look and let me know what you think.

eernstg reviewed (approved):
This is great! (and I learned a lot ;-)

eernstg commented:
Oh, one more thing: Wouldn't generic function instantiations have the same properties as function literal expressions, in the sense that it would be sound to de-couple the actual evaluation order from the type inference visiting order? Is it already true that the actual type arguments used to perform a generic function instantiation are capable of receiving information from other arguments? Otherwise it seems likely that we could treat generic function instantiations in a way that resembles the treatment of function literal expressions.

eernstg commented:
I forgot -- one more thing! I still don't understand the direction of the dependency edges.

As described (around lines 290-306), two nodes (each node is an actual argument) A and B have a dependency if certain criteria are satisfied, and as described the edge goes from A to B. However, the information flow goes from B to A, because it is B that may contain information after inference which is useful during inference on A. So we want to analyze B before A.

But with the graph {B} ⇐ {A} ⇒ {C, D}, it is stated (line 362) that the first stage consists of {A} and the second one consists of {B,C,D}.

I would expect the first (zeroth) stage to be {B,C,D} and the next one {A}.

eernstg commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,485 @@
+
+_In this example, that means that there will be two stages.  The first stage will
+consist of argument A, and the second stage will consist of arguments B, C, and
+D._
```
I'd expect the opposite order.

As described (around lines 290-306), two nodes (each node is an actual argument) A and B have a dependency if certain criteria are satisfied, and as described the edge goes from A to B. However, the information flow goes from B to A, because it is B that may contain information after inference which is useful during inference on A. So we want to analyze B before A.

But with the graph {B} ⇐ {A} ⇒ {C, D}, it is stated (line 362) that the first stage consists of {A} and the second one consists of {B,C,D}.

I would expect the first (zeroth) stage to be {B,C,D} and the next one {A}.

stereotype441 replied:
You're right.  Fixed.

stereotype441 commented:
> Oh, one more thing: Wouldn't generic function instantiations have the same properties as function literal expressions, in the sense that it would be sound to de-couple the actual evaluation order from the type inference visiting order? Is it already true that the actual type arguments used to perform a generic function instantiation are capable of receiving information from other arguments? Otherwise it seems likely that we could treat generic function instantiations in a way that resembles the treatment of function literal expressions.
> 
> Edit: Perhaps generic function instantiation takes place during the phase 'type checking', which means that information from other arguments _do_ get to flow into the generic function instantiation procedure. That would fit in with the behavior I can see when trying out an example like this one:
> 
> ```dart
> X f<X extends num>(X x1, X x2) => x1 < x2 ? x2 : x1;
> 
> void main() {
>   Iterable<int> values = [1, 6, 2];
>   var largestValue = values.fold(0, f);
> }
> ```
> 
> So this would imply that there is no need to treat generic function instantiation the same as function literal expressions.

Yes, I believe you are exactly right.

leafpetersen commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,485 @@
+     corresponding parameter type from the target function type.  This produces
+     additional type constraints beyond those gathered in step 3.
+
+   * Horizontal inference: if generic inference is needed, and this is not the
```
This doesn't make clear (per my previous comment) whether we "freeze" the results here or not (presumably we do).  That is, if we finish stage `k` with the constraint `X <: num`, do we finalize `X` as `num`, or do we allow stage `k+1` to further refine the solution for `X` (e.g. by generating a constraint `X <: int`)?  I suspect we want to "freeze" the solution, unfortunately, because otherwise different levels of inference may see different solutions for `X`.  Though maybe that's ok?  It's a little confusing, but then... inference is confusing.  :)

stereotype441 replied:
Good point.  In https://github.com/dart-lang/language/pull/2236 I've defined precisely when and how my implementation "freezes" the results of type inference.  Hopefully it's clear in that PR that we freeze the type inference of any type parameter as soon as we infer a type schema for it that doesn't contain `_`.

leafpetersen reviewed (commented):
A couple of comments/questions, but this is looking good.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,485 @@
+Today this doesn't work, because without the leading `int`, the downwards
+inference phase of type inference has no information with which to choose a
+preliminary type for the type parameter `T`, so the type context used for
+inference of the function literal `(a, b) => a < b ? b : a` is `? Function(?,
```
Addressed in https://github.com/dart-lang/language/pull/2236.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,485 @@
+  (denoted `_`), which allows representation of incomplete type information.
+
+- An expression's "type context" is a type schema representing what is known
+  about the probable type of the expression just before the portion of the type
```
I clarified this in https://github.com/dart-lang/language/pull/2236.  I avoided the term "required", but I clarified that in most circumstances the context represents the set of types that would be accepted without error, and I gave an example of one circumstance where it doesn't (demotion).

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,485 @@
+  possible values that may be assigned to an implicit type argument.
+
+- "Subtype constraint generation" (defined [here][subtype constraint
+  generation]) is an operation on two type schemas that produces a set of type
```
I see that your suggested rephrasing is more precise, but personally, I find it harder to grasp quickly.  Given that this is meant to be a reference to a definition in another doc, rather than a normative definition in and of itself, my preference is to leave it as is.

stereotype441 commented on working/0731 - horizontal inference/feature-specification.md:
```diff
@@ -0,0 +1,485 @@
+   invocation's type context.  This produces an initial set of type constraints.
+   Then, using those constraints, find the constraint solution for the target
+   function type's type variables, considering bounds.  This produces a
+   preliminary mapping of type variables to type schemas.
```
https://github.com/dart-lang/language/pull/2236 adds a definititive description of how we freeze type variables, split between this doc and `resources/type-system/inference.md`.
